politeiawww/mail/client.go

```diff
@@ -0,0 +1,293 @@
+	// new email history.
+	var (
+		valid     = make([]string, 0, len(users))
+		warning   = make([]string, 0, len(users))
+		histories = make(map[uuid.UUID]user.EmailHistory, len(users))
+	)
+	for userID, email := range users {
+		history, ok := hs[userID]
+		if !ok {
+			// User does not have a mail history yet, add user to valid
+			// recipients and create his email history.
+			histories[userID] = user.EmailHistory{
+				Timestamps:       []int64{time.Now().Unix()},
+				LimitWarningSent: false,
+			}
+			valid = append(valid, email)
+			continue
+		}
+
+		// Filter timestamps for the past rate limit period.
+		history.Timestamps = filterTimestamps(history.Timestamps,
+			c.rateLimitPeriod)
+
+		// Check if user can receive the email notification.
+		if len(history.Timestamps) < c.rateLimit {
+			// Rate limit has not been hit, add user to valid recipients and
+			// update email history.
+			valid = append(valid, email)
+			history.Timestamps = append(history.Timestamps, time.Now().Unix())
+			history.LimitWarningSent = false
+			histories[userID] = history
+		}
+
+		// Check if user has hit the email rate limit and needs to be warned.
+		if len(history.Timestamps) == c.rateLimit && !history.LimitWarningSent {
+			// Rate limit has been hit with the last email notification above.
+			// If limit warning email has not yet been sent, add user to
+			// warning recipients and update email history.
+			warning = append(warning, email)
+			history.LimitWarningSent = true
+			histories[userID] = history
+		}
+	}
+
+	return &filteredRecipients{
+		valid:     valid,
+		warning:   warning,
+		histories: histories,
+	}, nil
+}
+
+// filterTimestamps filters out timestamps from the passed in slice that comes
+// before the specified delta time duration.
+func filterTimestamps(in []int64, delta time.Duration) []int64 {
+	before := time.Now().Add(-delta)
+	out := make([]int64, 0, len(in))
+
+	for _, ts := range in {
+		timestamp := time.Unix(ts, 0)
+		if timestamp.Before(before) {
+			continue
+		}
+		out = append(out, ts)
+	}
+
+	return out
+}
+
+// The limit email is sent to users as a warning when they hit the email rate
+// limit.
+const limitEmailSubject = "Email Rate Limit Hit"
+const limitEmailBody = `
+Your email rate limit for the past 24 hours has been hit. This measure is used to avoid malicious users from spamming Politeia's email server. You will not receive any notification emails for 24 hours.
+
+We apologize for any inconvenience.
+`
+
+// NewClient returns a new client.
+func NewClient(host, user, password, emailAddress, certPath string, skipVerify bool, rateLimit int, db user.MailerDB) (*client, error) {
+	// Email is considered disabled if any of the required user
+	// credentials are missing.
+	if host == "" || user == "" || password == "" {
+		log.Infof("Mail: DISABLED")
+		return &client{
+			disabled: true,
+		}, nil
+	}
+
+	// Parse mail host
+	h := fmt.Sprintf("smtps://%v:%v@%v", user, password, host)
```

Comment:
This needs to be sanitized with a regex.

Reply:
Are you ok with the host being verified using `url.Parse()`?  What kind of sanitization do you want on the username and password?

Reply:
Below we `url.Parse(h)` for the full host built with `h:= fmt.Sprintf(...)`. Is that enough?

Reply:
Yeah, I am ok with that. Looked at the function and it is pretty thorough.